src/timely-util/src/operators_async_ext.rs

```diff
@@ -166,24 +178,49 @@ impl<G: Scope> OperatorBuilderExt<G> for OperatorBuilder<G> {
                 if had_pending_notify && !scheduler.is_notified() {
                     waker.wake_by_ref();
                 }
+
+                operator_incomplete
             }
         });
     }
 }
 
-#[allow(unused_macros)]
+use differential_dataflow::lattice::Lattice;
+#[doc(hidden)]
+pub fn multi_meet<T: Lattice + Clone>(antichains: &[Antichain<T>]) -> Antichain<T> {
+    antichains
+        .iter()
+        .fold(Antichain::new(), |acc, x| acc.meet(&x))
+}
+
+/// Convenience macro used to wrap what might otherwise be the argument to `build_reschedule`.
+#[macro_export]
 macro_rules! async_op {
-    (|$capabilities:ident, $frontiers:ident| $body:block) => {
-        move |mut capabilities, mut frontiers, scheduler| async move {
+    (|$capabilities:ident, $frontier:ident| $body:block) => {
+        move |mut capabilities, frontiers, scheduler| async move {
             loop {
                 scheduler.notified().await;
-                // rebind to mutable references to make sure they can't be accidentally dropped
+
+                // We simplify into one single Antichain because no current users of `async_op`
+                // require treating multiple input frontiers separately and this is the easiest
+                // way we can ensure that the `$body` does not hold a borrow to the frontier at
+                // all times
+                //
+                // Pull out to function so we don't require import from differential_dataflow for
+                // user of macro.
+                let __frontier = $crate::operators_async_ext::multi_meet(&frontiers.borrow());
+                let $frontier = __frontier.borrow();
                 #[allow(unused_mut)]
                 let mut $capabilities = &mut capabilities;
-                #[allow(unused_mut)]
-                let mut $frontiers = &mut frontiers;
-                let _: () = async { $body }.await;
+
+                if !async { $body }.await && $frontier.is_empty() {
+                    break;
+                }
             }
+
+            // Make sure they can't be accidentally dropped in `$body`
+            drop(capabilities);
```

Comment:
This isn't necessary because the `capabilities` binding that `$body` has access to is a fresh mutable borrow that is created on every iteration of the loop (the `let mut $capabilities = &mut capabilities;` line). The same was true for the frontiers in the previous version of the code